### ISSUE TITLE
docker.errors.DockerException while running beobench run --config config.yaml

### ISSUE THREAD

**Issue by HYDesmondLiu** [rdnfn/beobench#82]
First, thank you for sharing the codes. 
I was trying to run examples provided in the readme section. 
`beobench run --config config.yaml` using provided config.yaml and agent.py
   
However, I got this `docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', PermissionError(13, 'Permission denied'))` error.

Please let me know if any further information is needed to debug.

**Environments:**
Python 3.8.5
Ubuntu 18.04.3 LTS

**Detailed message:**
```
`Beobench ⚡️Starting experiment run ...
Beobench ⚡️Running experiment in container with environment MixedUseFanFCU-v0 and agent from ./agent.py. Sample 1 of 1.
Beobench ⚡️Recognised integration named sinergym.
Traceback (most recent call last):
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 950, in send
    self.connect()
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/transport/unixconn.py", line 30, in connect
    sock.connect(self.unix_socket)
PermissionError: [Errno 13] Permission denied

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/connectionpool.py", line 724, in urlopen
    retries = retries.increment(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/util/retry.py", line 403, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/home/hsinyu/anaconda3/lib/python3.8/http/client.py", line 950, in send
    self.connect()
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/transport/unixconn.py", line 30, in connect
    sock.connect(self.unix_socket)
urllib3.exceptions.ProtocolError: ('Connection aborted.', PermissionError(13, 'Permission denied'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/api/client.py", line 214, in _retrieve_server_version
    return self.version(api_version=False)["ApiVersion"]
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/api/daemon.py", line 181, in version
    return self._result(self._get(url), json=True)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/utils/decorators.py", line 46, in inner
    return f(self, *args, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/api/client.py", line 237, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', PermissionError(13, 'Permission denied'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hsinyu/anaconda3/bin/beobench", line 8, in <module>
    sys.exit(cli())
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/beobench/cli.py", line 122, in run
    beobench.experiment.scheduler.run(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/beobench/experiment/scheduler.py", line 172, in run
    _build_and_run_in_container(config)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/beobench/experiment/scheduler.py", line 202, in _build_and_run_in_container
    image_tag = beobench.experiment.containers.build_experiment_container(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/beobench/experiment/containers.py", line 91, in build_experiment_container
    if not force_build and check_image_exists(stage2_image_tag):
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/beobench/experiment/containers.py", line 23, in check_image_exists
    client = docker.from_env()
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/client.py", line 96, in from_env
    return cls(
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/client.py", line 45, in __init__
    self.api = APIClient(*args, **kwargs)
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/api/client.py", line 197, in __init__
    self._version = self._retrieve_server_version()
  File "/home/hsinyu/anaconda3/lib/python3.8/site-packages/docker/api/client.py", line 221, in _retrieve_server_version
    raise DockerException(
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', PermissionError(13, 'Permission denied'))`
```

**Comment by HYDesmondLiu** [comment]
@rdnfn 
Thank you for the prompt reply. Previously indeed I was not in the docker user group.
However, after being added to the group. I got another error.

And, I use the same `agent.py` and `config.yaml`as the examples provided in the readme page. The env. name is `Eplus-5Zone-hot-continuous-v1`. I am not sure where that environment is set.

<img width="1601" alt="Screen Shot 2022-06-30 at 9 46 14 AM" src="https://user-images.githubusercontent.com/43663175/176574995-9b4e0878-8aec-46ad-91c8-5468d8f7a13d.png">
<img width="1564" alt="Screen Shot 2022-06-30 at 9 49 24 AM" src="https://user-images.githubusercontent.com/43663175/176575374-737afd31-0127-46d0-98fa-7e24c404d09e.png">

**Comment by rdnfn** [comment]
@HYDesmondLiu Glad to hear that the first error is fixed. 

And you're absolutely right, the wrong env name in the log is not your fault but rather a bug in the logging code in Beobench (wouldn't affect your experiment itself though). Sorry about the confusion. I fixed it already in the `dev/general` branch and that fix will be added in the next release.

About the docker `unkown shorthand flag` error: I am having a difficult time to reproduce it. Would you be able to share your docker version? You can find out with the command `docker --version`. Thanks!

**Comment by HYDesmondLiu** [comment]
@rdnfn 
Thanks for the prompt reply.
My docker version is listed below:
`Docker version 19.03.12, build 48a66213fe`

**Comment by rdnfn** [comment]
Thanks for your patience and sharing the version! Good news, I have been able to reproduce the error now and I also hopefully have been able to fix it. Please update to the latest (new) Beobench version (v0.5.2) using `pip install beobench --upgrade`. Let me know if this update resolved your problem!

**Background:** The problem appears to be that your version of docker (v19.03) does not support the `docker buildx` subcommand of docker that Beobench (v0.5.1) uses by default to build experiment container images. This problem is not very easily visible because for some reason it gets hidden behind this `unknown flag` error (see output below). As buildx is not required for all use-cases, I have disabled it now where possible (if not on ARM64 architecture).

For future reference, output from the tests I ran:

```
/ # docker --version
Docker version 19.03.15, build 99e3ed8
/ # docker buildx build
docker: 'buildx' is not a docker command.
See 'docker --help'
/ # docker buildx build -t
unknown shorthand flag: 't' in -t
See 'docker --help'.

Usage:  docker [OPTIONS] COMMAND

A self-sufficient runtime for containers

Options:
      --config string      Location of client config files (default "/root/.docker")

	<more help output>

```

**Comment by HYDesmondLiu** [comment]
Hi @rdnfn,
    Thanks for the quick fix, after upgrading to `beobench v0.5.2`, while running the same command `beobench run --config config.yaml`. I have got another error as shown in the snapshot below, something problems related to pip install.  

<img width="1558" alt="Screen Shot 2022-07-02 at 8 26 04 AM" src="https://user-images.githubusercontent.com/43663175/176979776-822c462b-86f4-406b-8887-bd692841a423.png">

**Comment by david-woelfle** [comment]
> Hi @rdnfn, Thanks for the quick fix, after upgrading to `beobench v0.5.2`, while running the same command `beobench run --config config.yaml`. I have got another error as shown in the snapshot below, something problems related to pip install.
> 
> <img alt="Screen Shot 2022-07-02 at 8 26 04 AM" width="1558" src="https://user-images.githubusercontent.com/43663175/176979776-822c462b-86f4-406b-8887-bd692841a423.png">

Can confirm, same problem here.

**Comment by rdnfn** [comment]
@david-woelfle and @HYDesmondLiu thanks for raising this problem! I am working on it ... will release a fix soon.

**Comment by rdnfn** [comment]
Thanks again @HYDesmondLiu and @david-woelfle for finding this error! Thanks for your patience!

I have now updated the development version of Beobench with a fix for this issue. You can install the latest development version using the command:

```
pip install git+https://github.com/rdnfn/beobench.git@dev/general
```

If you could try to install this and let me know if this resolves your error on your machines? If that's the case, I will publish a new version (v0.5.3) with this fix. Thanks so much for your help making Beobench better!

**Note:**
The reason the fix took me so long is because I ran into another unrelated bug in the GitHub CI related to the use of Beobench inside docker-in-docker containers. I think this is unlikely, but if you're using this kind of setup (dind) then have look at #85.

**Background:**
I used a form of conditional statement inside the Dockerfile, but this appears to have broken without `buildx`. Thus, I moved this logic directly into Python.

**Comment by HYDesmondLiu** [comment]
Hi @rdnfn,
    Thanks for the quick fix. this version works for me.
However, are you planning to add more content to [advanced usage page](https://beobench.readthedocs.io/en/latest/advanced_usage.html)?
    I have some question about advanced usage, for example:
1. I tried to invoke my own algorithm in the agent script, however, it cannot recognize modules in the same directory.
2. Also, the first yaml configuration in [the page](https://beobench.readthedocs.io/en/latest/guides/configuration.html) with _Energym_ does not work. 
    If I run the yaml, it shows error messages as this:
<img width="1071" alt="Screen Shot 2022-07-07 at 10 07 22 PM" src="https://user-images.githubusercontent.com/43663175/177921006-4f6c5ccd-0284-4acd-acea-60a076fe7da7.png">
   If I use pdb to debug it shows a similar error as the original bug reported.    
<img width="1568" alt="Screen Shot 2022-07-07 at 10 08 36 PM" src="https://user-images.githubusercontent.com/43663175/177921119-f06771bc-33a6-4196-9663-648c5242d6e1.png">

**Comment by rdnfn** [comment]
Thanks for testing this @HYDesmondLiu! Glad to hear this works now.

About your other points:

> However, are you planning to add more content to advanced usage page?

Absolutely, this is work in progress!

> 1. I tried to invoke my own algorithm in the agent script, however, it cannot recognize modules in the same directory.

This is unfortunately not supported yet, but I am hoping to add this feature soon. #70 tracks the mounting of the directory of the agent script (e.g. access to files in the same folder as the agent script and its subfolders). #71 tracks the installation of agent script specific external pypi dependencies (e.g. pip install packagename).

> 2. Also, the first yaml configuration in the page with Energym does not work. If I run the yaml, it shows error messages as this:

Thanks for flagging this. This error is likely because of an env.config setting. Would you be able to open another issue for this problem with a copy of the exact YAML file that you used?

With that I will close this specific issue now, the changes will be included in the next release (v0.5.3). Feel free to open a new issues for any other problems or questions you run into!

**Comment by HYDesmondLiu** [comment]
@rdnfn 
    Thanks for the detailed responses. I am glad these are considered in recent revision soon.
    Look forward to the revision.
